src/main/java/org/opentripplanner/routing/stoptimes/StopTimesHelper.java

```diff
@@ -111,32 +95,30 @@ public static List<StopTimesInPattern> stopTimesForStop(
   ) {
     List<StopTimesInPattern> ret = new ArrayList<>();
 
-    Collection<TripPattern> patternsForStop = transitService.getPatternsForStop(stop, true);
-    for (TripPattern pattern : patternsForStop) {
+    var servicesRunning = transitService.getServicesRunningForDate(serviceDate);
+    Instant midnight = ServiceDateUtils
+      .asStartOfService(serviceDate, transitService.getTimeZone())
+      .toInstant();
+
+    for (TripPattern pattern : transitService.getPatternsForStop(stop, true)) {
       StopTimesInPattern stopTimes = new StopTimesInPattern(pattern);
-      Timetable tt;
-      TimetableSnapshot timetableSnapshot = transitService.getTimetableSnapshot();
-      if (timetableSnapshot != null) {
-        tt = timetableSnapshot.resolve(pattern, serviceDate);
-      } else {
-        tt = pattern.getScheduledTimetable();
-      }
-      var servicesRunning = transitService.getServicesRunningForDate(serviceDate);
-      Instant midnight = ServiceDateUtils
-        .asStartOfService(serviceDate, transitService.getTimeZone())
-        .toInstant();
-      int sidx = 0;
-      for (var currStop : pattern.getStops()) {
+      Timetable tt = Optional
+        .ofNullable(transitService.getTimetableSnapshot())
+        .map(timetableSnapshot -> timetableSnapshot.resolve(pattern, serviceDate))
+        .orElse(pattern.getScheduledTimetable());
```

Comment:
This logic should be pushed into TransitService, something like this:
```
Timetable tt = transitService.findTimetable(pattern, serviceDate);
```

Reply:
We should probably do some analysis/design here. The `TransitService` should be request scoped, so passing the TimeTableSnapshot down the call stack is not necessary. Maybe it is possible to hide the TimeTableSnapshot form the APIs, I am not sure.

Reply:
The issue why it is being passed down, is that we want to always use the same snapshot for a single API call. We could initialize it lazily in `DefaultTransitService` when accessed for the first time. Unfortunately it is also used in the updaters, where we want it to be periodically refreshed. We need to be mindful of the lifecycle in order to avoid bugs like the one fixed in #4343

Reply:
I just noticed, that we have `lazyGetTimeTableSnapShot`, which we should use exactly for things like this.

Reply:
I removed `TimeTableSnapshot` from `TransitService` and expose individual methods instead

Reply:
Nice!